### PR TITLE
Cyclopedia Character: Send General Stats

### PR DESCRIPTION
**Pull request by omarcopires**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Credits to
@zbizu
@ArturKnopik 
@MillhioreBT
https://github.com/OTAcademy

### Changes Proposed
Add two new LUA functions:

```lua
self:getLevelPercent()
self:getMagicLevelPercent()
```
image

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: data/scripts/network/cyclopedia_character.lua [hunk 1]
```diff
@@ -24,10 +24,78 @@ local function sendBasicInfo(self, msg)
...
+
+	msg:addU16(foodTime)
+	msg:addU16(self:getOfflineTrainingTime() / 60 / 1000)
+	msg:addU16(self:getSpeed() / 2)
+	msg:addU16(self:getBaseSpeed() / 2)
+	msg:addU32(self:getCapacity()) -- cap + boost
+	msg:addU32(self:getCapacity())
+	msg:addU32(self:getFreeCapacity())
+
+	msg:addByte(8) -- ??
```
Comment (EPuncker) [inline data/scripts/network/cyclopedia_character.lua:68]: blessings?

File: data/scripts/network/cyclopedia_character.lua [hunk 2]
```diff
@@ -75,6 +75,87 @@ local function sendCombatStats(self, msg)
 	return true
 end
 
+local function getLevelPercent(player)
+	local currentLevelExp = Game.getExperienceForLevel(player:getLevel())
+	local nextLvlExp = Game.getExperienceForLevel(player:getLevel() +1)
```
Comment (EPuncker) [inline data/scripts/network/cyclopedia_character.lua:80]:
```suggestion
	local nextLvlExp = Game.getExperienceForLevel(player:getLevel() + 1)
```

**Review by EPuncker — APPROVED** [review]

File: data/scripts/network/cyclopedia_character.lua [hunk 3]
```diff
@@ -24,6 +24,87 @@ local function sendBasicInfo(self, msg)
...
+	[6] = CYCLOPEDIA_SKILL_FISHING
+}
+
+local function sendGeneralStats(self, msg)
+	msg:addU64(self:getExperience())
+	msg:addU16(self:getLevel())
+	msg:addByte(math.floor(getLevelPercent(self)))
+
+	msg:addU16(self:getClientExpDisplay())
+	msg:addU32(0) -- turnament exp (deprecated)
```
Comment (ranisalt) [inline data/scripts/network/cyclopedia_character.lua:50]:
```suggestion
	msg:addU32(0) -- tournament exp (deprecated)
```

File: data/scripts/network/cyclopedia_character.lua [hunk 4]
```diff
@@ -24,6 +24,87 @@ local function sendBasicInfo(self, msg)
...
+
+	msg:addByte(8) -- ??
+	msg:addByte(1) -- ??
+
+	-- magic level
+	msg:addU16(self:getMagicLevel())
+	msg:addU16(self:getBaseMagicLevel())
+
+	-- loyalty bonus
+	msg:addU16(self:getBaseMagicLevel())
```
Comment (ranisalt) [inline data/scripts/network/cyclopedia_character.lua:88]: Base magic level twice?

File: data/scripts/network/cyclopedia_character.lua [hunk 5]
```diff
@@ -24,6 +24,87 @@ local function sendBasicInfo(self, msg)
...
+	if condition then
+		foodTime = condition:getTicks() /  1000
+	end
+
+	msg:addU16(foodTime)
+	msg:addU16(self:getOfflineTrainingTime() / 60 / 1000)
+	msg:addU16(self:getSpeed() / 2)
+	msg:addU16(self:getBaseSpeed() / 2)
+
+	local infiniteCapacity = self:hasFlag(PlayerFlag_HasInfiniteCapacity) and 1000000
```
Comment (ranisalt) [inline data/scripts/network/cyclopedia_character.lua:75]: Does `1000000` show a special value, or is this just an arbitrarily large value?

File: data/scripts/network/cyclopedia_character.lua [hunk 6]
```diff
@@ -24,6 +24,87 @@ local function sendBasicInfo(self, msg)
...
+	msg:addU16(self:getMaxHealth())
+	msg:addU16(self:getMana())
+	msg:addU16(self:getMaxMana())
+	msg:addByte(self:getSoul())
+	msg:addU16(self:getStamina())
+
+	local foodTime = 0
+	local condition = self:getCondition(CONDITION_REGENERATION, CONDITIONID_DEFAULT)
+	if condition then
+		foodTime = condition:getTicks() /  1000
```
Comment (ranisalt) [inline data/scripts/network/cyclopedia_character.lua:67]:
```suggestion
		foodTime = condition:getTicks() / 1000
```

File: data/scripts/network/cyclopedia_character.lua (hunk 5 above)
Comment (omarcopires) [reply data/scripts/network/cyclopedia_character.lua:75]: it's the same value that's defined in the engine, anyway it's just visual for admins xD

File: data/scripts/network/cyclopedia_character.lua (hunk 4 above)
Comment (omarcopires) [reply data/scripts/network/cyclopedia_character.lua:85]: it's the magic level information and includes the loyalty bonus, but it's confusing to understand

File: data/scripts/network/cyclopedia_character.lua (hunk 5 above)
Comment (ranisalt) [reply data/scripts/network/cyclopedia_character.lua:75]: If the client uses this, then it's fine, I would just assign it to a name `INFINITE_CAPACITY`

File: data/scripts/network/cyclopedia_character.lua (hunk 4 above)
Comment (omarcopires) [reply data/scripts/network/cyclopedia_character.lua]: I don't think I have enough knowledge xD

**Review by EPuncker — APPROVED** [review]

File: data/scripts/network/cyclopedia_character.lua [hunk 7]
```diff
@@ -24,6 +24,77 @@ local function sendBasicInfo(self, msg)
...
+		msg:addByte(clientSkillsId[i])
+		msg:addU16(self:getEffectiveSkillLevel(i))
+		msg:addU16(self:getSkillLevel(i))
+
+		-- base + loyalty bonus
+		msg:addU16(self:getSkillLevel(i))
+		msg:addU16(self:getSkillPercent(i) * 100)
+	end
+
+	msg:addByte(0) -- magic boost (elemend and value)
```
Comment (EPuncker) [inline data/scripts/network/cyclopedia_character.lua:91]:
```suggestion
	msg:addByte(0) -- magic boost (element and value)
```

**Review by EPuncker — APPROVED** [review]

File: data/lib/core/constants.lua [hunk 8]
```diff
@@ -15,6 +15,15 @@ VOCATION_ELDER_DRUID = 6
...
 VOCATION_ELITE_KNIGHT = 8
 
+CYCLOPEDIA_SKILL_MAGIC = 1
+CYCLOPEDIA_SKILL_SHIELDING = 6
+CYCLOPEDIA_SKILL_DISTANCE = 7
+CYCLOPEDIA_SKILL_SWORD = 8
+CYCLOPEDIA_SKILL_CLUB = 9
+CYCLOPEDIA_SKILL_AXE = 10
+CYCLOPEDIA_SKILL_FIST = 11
+CYCLOPEDIA_SKILL_FISHING = 13
```
Comment (ranisalt) [inline data/lib/core/constants.lua:25]: Do we have any idea what is inbetween i.e. what is 2-5 and 12?

Comment (omarcopires) [reply data/lib/core/constants.lua:25]: i have no idea what they are, i just know that the current mapping works as it should, in fact all the code works as expected.

**Pull request closed without merging**